File: Editorial/6_Zigzag_Conversion/TypeScript/1_Simulate_Zig_Zag_Movement.ts
function convert(s: string, numRows: number): string {
    if (numRows === 1) {
        return s;
    }

    let n = s.length;
    let sections = Math.ceil(n / (2 * numRows - 2));
    let numCols = sections * (numRows - 1);

    let matrix = Array.from({ length: numRows }, () =>
        Array(numCols).fill(" "),
    );

    let currRow = 0,
        currCol = 0;
    let currStringIndex = 0;

    // Iterate in zig-zag pattern on matrix and fill it with string characters.
    while (currStringIndex < n) {
        // Move down.
        while (currRow < numRows && currStringIndex < n) {
            matrix[currRow][currCol] = s[currStringIndex];
            currRow++;
            currStringIndex++;
        }

        currRow -= 2;
        currCol++;

        // Move up (with moving right also).
        while (currRow > 0 && currCol < numCols && currStringIndex < n) {
            matrix[currRow][currCol] = s[currStringIndex];
            currRow--;
            currCol++;
            currStringIndex++;
        }
    }

    let answer = matrix.map((row) => row.join(""));
    return answer.join("").split(" ").join("");
}

File: Editorial/6_Zigzag_Conversion/CSharp/1_Simulate_Zig_Zag_Movement.cs
public class Solution {
    public string Convert(string s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        int n = s.Length;
        int sections = (int)Math.Ceiling(n / (2 * numRows - 2.0));
        int numCols = sections * (numRows - 1);

        char[][] matrix = new char [numRows][];
        for (int i = 0; i < numRows; i++) {
            matrix[i] = new char[numCols];
            matrix[i] = Enumerable.Repeat(' ', numCols).ToArray();
        }

        int currRow = 0, currCol = 0;
        int currStringIndex = 0;

        // Iterate in zig-zag pattern on matrix and fill it with string
        // characters.
        while (currStringIndex < n) {
            // Move down.
            while (currRow < numRows && currStringIndex < n) {
                matrix[currRow][currCol] = s[currStringIndex];
                currRow++;
                currStringIndex++;
            }

            currRow -= 2;
            currCol++;

            // Move up (with moving right also).
            while (currRow > 0 && currCol < numCols && currStringIndex < n) {
                matrix[currRow][currCol] = s[currStringIndex];
                currRow--;
                currCol++;
                currStringIndex++;
            }
        }

        string answer = "";
        foreach (char[] row in matrix) {
            string rowStr = new string(row).Replace(" ", "");
            answer += rowStr;
        }

        return answer;
    }
}
